### PR TITLE
sub-menu

### PR DESCRIPTION
**Pull request by clementbalestrat**

### CODE REVIEW

File: sections/shared/Layout/SideNav/SideNav.tsx [hunk 1]
```diff
@@ -20,17 +21,25 @@ import { CryptoCurrency, Synths } from 'constants/currency';
 import { SIDE_NAV_WIDTH, zIndex } from 'constants/ui';
 
 import { MENU_LINKS, MIGRATE_MENU_LINKS } from '../constants';
+import SubMenu from './SubMenu';
 import PriceItem from './PriceItem';
 import PeriodBarStats from './PeriodBarStats';
 import BalanceItem from './BalanceItem';
 import CRatioBarStats from './CRatioBarStats';
 
+const getKeyValue = <T extends object, U extends keyof T>(obj: T) => (key: U) => obj[key];
```
Comment (clementbalestrat) [inline sections/shared/Layout/SideNav/SideNav.tsx:30]: To fix a TS error with objects and indices.

File: sections/shared/Layout/SideNav/SideNav.tsx [hunk 2]
```diff
@@ -20,17 +21,25 @@ import { CryptoCurrency, Synths } from 'constants/currency';
...
+
 const SideNav: FC = () => {
 	const { t } = useTranslation();
 	const { asPath } = useRouter();
+	const menuLinkItemRefs = useRef({});
```
Comment (clementbalestrat) [inline sections/shared/Layout/SideNav/SideNav.tsx:35]: Using refs here to catch the `y` position of every menu item, so we can then pass it to a props and use `top: y px` so menu and sub menu can be aligned

File: sections/shared/Layout/SideNav/SideNav.tsx [hunk 3]
```diff
@@ -20,17 +21,25 @@ import { CryptoCurrency, Synths } from 'constants/currency';
...
 	const SNX24hrPricesQuery = useSNX24hrPricesQuery();
 	const cryptoBalances = useCryptoBalances();
 	const synthsBalancesQuery = useSynthsBalancesQuery();
+	const [subMenuConfiguration, setSubMenuConfiguration] = useState({
```
Comment (clementbalestrat) [inline sections/shared/Layout/SideNav/SideNav.tsx:39]: As explained above, a subMenu configuration is passed to `SubMenu` with two infos: 
- and array with all the routes
- the `y` position it should be displayed

File: sections/shared/Layout/SideNav/SideNav.tsx [hunk 4]
```diff
@@ -51,22 +60,46 @@ const SideNav: FC = () => {
 	const menuLinks = totalBalancePendingMigration > 0 ? MIGRATE_MENU_LINKS : MENU_LINKS;
 
 	return (
-		<SideNavContainer>
+		<SideNavContainer
```
Comment (clementbalestrat) [inline sections/shared/Layout/SideNav/SideNav.tsx:63]: if the cursor leaves the container then we hide the SubMenu

File: sections/shared/Layout/SideNav/SideNav.tsx [hunk 5]
```diff
@@ -51,22 +60,46 @@ const SideNav: FC = () => {
...
+					<Svg src={StakingLogo} />
 				</Link>
 			</StakingLogoWrap>
 			<MenuLinks>
-				{menuLinks.map(({ i18nLabel, link }) => (
+				{menuLinks.map(({ i18nLabel, link, subMenu }, i) => (
 					<MenuLinkItem
+						ref={(r) => {
+							if (subMenu) {
+								menuLinkItemRefs.current = { ...menuLinkItemRefs.current, [i]: r };
```
Comment (clementbalestrat) [inline sections/shared/Layout/SideNav/SideNav.tsx:76]: For every `MenuItem` which contain `subMenu` we save their ref to `menuLinkItemRefs`

File: sections/shared/Layout/SideNav/SideNav.tsx [hunk 6]
```diff
@@ -51,22 +60,46 @@ const SideNav: FC = () => {
...
+							}
+						}}
+						onMouseEnter={() => {
+							setSubMenuConfiguration(
+								subMenu
+									? {
+											routes: subMenu as any,
```
Comment (clementbalestrat) [inline sections/shared/Layout/SideNav/SideNav.tsx:83]: When a MenuItem is hovered we get its `boundingClientRect().y` and save it to the `subMenuConfiguration` object

File: sections/shared/Layout/SideNav/SubMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+type SubMenuProps = {
+	config: SubMenuConfiguration;
+	currentPath: string;
+};
+
+const SubMenu: FC<SubMenuProps> = ({ config, currentPath }) => {
+	const { t } = useTranslation();
+	return typeof document !== 'undefined'
+		? createPortal(
```
Comment (clementbalestrat) [inline sections/shared/Layout/SideNav/SubMenu.tsx:25]: Using Portal here as the SideNav has a `fixed` position, which means there is no way to put the subMenu below with `z-index`. The only way is to send this component to the root.

File: sections/shared/Layout/SideNav/SubMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+				<Container isVisible={!!config?.routes ?? false}>
+					<Inner>
+						<FloatingContent topPosition={config?.topPosition ?? 0}>
+							<LinkList>
+								{config?.routes?.map(({ i18nLabel, subLink }) => (
+									<MenuLinkItem isActive={currentPath === subLink}>
```
Comment (evgenyboxer) [inline sections/shared/Layout/SideNav/SubMenu.tsx:31]: Missing `key` right?

Comment (clementbalestrat) [reply sections/shared/Layout/SideNav/SubMenu.tsx:31]: yep good find

File: sections/shared/Layout/SideNav/SubMenu.tsx (hunk 7 above)
Comment (evgenyboxer) [reply sections/shared/Layout/SideNav/SubMenu.tsx:25]: mice!

File: sections/shared/Layout/SideNav/SideNav.tsx (hunk 1 above)
Comment (evgenyboxer) [reply sections/shared/Layout/SideNav/SideNav.tsx:30]: I wonder if it's because `menuLinkItemRefs` and `subMenuConfiguration` need to be typed

**Review by evgenyboxer — APPROVED** [review]
@clementbalestrat great stuff!

I think typing the ref/state might help with the TS errors, can be fixed later.

File: sections/shared/Layout/SideNav/SideNav.tsx (hunk 1 above)
Comment (clementbalestrat) [reply sections/shared/Layout/SideNav/SideNav.tsx:30]: Yeah I think so too.

**Pull request merged**